pkg/sfu/buffer/rtpstats.go

```diff
@@ -371,6 +378,75 @@ func (r *RTPStats) Update(rtph *rtp.Header, payloadSize int, paddingSize int, pa
 		return
 	}
 
+	if r.resyncOnNextPacket {
+		r.resyncOnNextPacket = false
+
+		if r.initialized {
+			extHighestSN := r.sequenceNumber.GetExtendedHighest()
+			var newestPacketCount uint64
+			var paddingOnlyDrops uint64
+			var extExpectedHighestSN uint64
+			var expectedHighestSN uint16
+			var snCycles uint64
+
+			extHighestTS := r.timestamp.GetExtendedHighest()
+			var newestTS uint64
+			var extExpectedHighestTS uint64
+			var expectedHighestTS uint32
+			var tsCycles uint64
+			if r.srNewest != nil {
+				newestPacketCount = r.srNewest.PacketCountExt
+				paddingOnlyDrops = r.srNewest.PaddingOnlyDrops
+				if newestPacketCount != 0 {
+					extExpectedHighestSN = r.sequenceNumber.GetExtendedStart() + newestPacketCount
+					if r.shouldDiscountPaddingOnlyDrops {
+						extExpectedHighestSN -= paddingOnlyDrops
+					}
+					expectedHighestSN = uint16(extExpectedHighestSN & 0xFFFF)
+					snCycles = extExpectedHighestSN & 0xFFFF_FFFF_FFFF_0000
+					if rtph.SequenceNumber-expectedHighestSN < (1<<15) && rtph.SequenceNumber < expectedHighestSN {
+						snCycles += (1 << 16)
+					}
+					if snCycles != 0 && expectedHighestSN-rtph.SequenceNumber < (1<<15) && expectedHighestSN < rtph.SequenceNumber {
+						snCycles -= (1 << 16)
+					}
+				}
+
+				newestTS = r.srNewest.RTPTimestampExt
+				extExpectedHighestTS = newestTS
+				expectedHighestTS = uint32(extExpectedHighestTS & 0xFFFF_FFFF)
+				tsCycles = extExpectedHighestTS & 0xFFFF_FFFF_0000_0000
+				if rtph.Timestamp-expectedHighestTS < (1<<31) && rtph.Timestamp < expectedHighestTS {
+					tsCycles += (1 << 32)
+				}
+				if tsCycles != 0 && expectedHighestTS-rtph.Timestamp < (1<<31) && expectedHighestTS < rtph.Timestamp {
+					tsCycles -= (1 << 32)
+				}
+			}
+			r.sequenceNumber.ResetHighest(snCycles + uint64(rtph.SequenceNumber) - 1)
+			r.timestamp.ResetHighest(tsCycles + uint64(rtph.Timestamp))
+			r.highestTime = packetTime
+			r.logger.Debugw(
+				"resync",
+				"newestPacketCount", newestPacketCount,
+				"paddingOnlyDrops", paddingOnlyDrops,
+				"extExpectedHighestSN", extExpectedHighestSN,
+				"expectedHighestSN", expectedHighestSN,
+				"snCycles", snCycles,
+				"rtpSN", rtph.SequenceNumber,
+				"beforeExtHighestSN", extHighestSN,
+				"afterExtHighestSN", r.sequenceNumber.GetExtendedHighest(),
+				"newestTS", newestTS,
+				"extExpectedHighestTS", extExpectedHighestTS,
+				"expectedHighestTS", expectedHighestTS,
+				"tsCycles", tsCycles,
+				"rtpTS", rtph.Timestamp,
+				"beforeExtHighestTS", extHighestTS,
+				"afterExtHighestTS", r.timestamp.GetExtendedHighest(),
+			)
+		}
+	}
```

Comment:
This calculation is not exact. There are probably very edge cases where if all stars line up, this could go wrong, but a very small chance. In the unlikely chance that happens, will affect A/V sync a bit I think. So, it is not bad.